src/input_handlers.h

```diff
@@ -0,0 +1,122 @@
+ * SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ */
+
+#ifndef INPUT_HANDLERS_H_
+#define INPUT_HANDLERS_H_
+
+#include "message_parser.h"
+
+class RecvFromInputHandler : public MessageParser<InPlaceAccumulation> {
+private:
+    SocketRecvData &m_recv_data;
+    uint8_t *m_actual_buf;
+    int m_actual_buf_size;
+public:
+    inline RecvFromInputHandler(Message *msg, SocketRecvData &recv_data):
+        MessageParser<InPlaceAccumulation>(msg),
+        m_recv_data(recv_data),
+        m_actual_buf_size(0)
+    {}
+
+    /** Receive pending data from a socket
+     * @param [in] socket descriptor
+     * @param [out] recvfrom_addr address to save peer address into
+     * @return status code
+     */
+    inline int receive_pending_data(int fd, struct sockaddr_in *recvfrom_addr)
+    {
+        int ret = 0;
+        socklen_t size = sizeof(struct sockaddr_in);
+        int flags = 0;
+        uint8_t *buf = m_recv_data.cur_addr + m_recv_data.cur_offset;
+
+/*
+    When writing onto a connection-oriented socket that has been shut down
+    (by the local or the remote end) SIGPIPE is sent to the writing process
+    and EPIPE is returned. The signal is not sent when the write call specified
+    the MSG_NOSIGNAL flag.
+    Note: another way is call signal (SIGPIPE,SIG_IGN);
+ */
+#ifndef WIN32
+        flags = MSG_NOSIGNAL;
+#endif
+
+#if defined(DEFINED_TLS)
+        if (g_fds_array[fd]->tls_handle) {
+            ret = tls_read(g_fds_array[fd]->tls_handle, buf, m_recv_data.cur_size);
+        } else
+#endif /* DEFINED_TLS */
+        {
+            ret = recvfrom(fd, buf, m_recv_data.cur_size,
+                    flags, (struct sockaddr *)recvfrom_addr, &size);
+        }
+        m_actual_buf = buf;
+        m_actual_buf_size = ret;
+
+#if defined(LOG_TRACE_MSG_IN) && (LOG_TRACE_MSG_IN == TRUE)
+        printf(">   ");
+        hexdump(buf, MsgHeader::EFFECTIVE_SIZE);
+#endif /* LOG_TRACE_MSG_IN */
+
+#if defined(LOG_TRACE_RECV) && (LOG_TRACE_RECV == TRUE)
+        LOG_TRACE("raw", "%s IP: %s:%d [fd=%d ret=%d] %s", __FUNCTION__,
+                  inet_ntoa(recvfrom_addr->sin_addr), ntohs(recvfrom_addr->sin_port), fd, ret,
+                  strerror(errno));
+#endif /* LOG_TRACE_RECV */
+
+        if (ret == 0 || errno == EPIPE || os_err_conn_reset()) {
+            /* If no messages are available to be received and the peer has performed an orderly
+             * shutdown,
+             * recv()/recvfrom() shall return 0
```

Comment:
Consider combining these two lines into one. Another section also has this comment `* shutdown,`.